File: src/Configuration/IniConfigFile.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

/// <summary>
/// Handles reading and writing INI-style configuration files.
/// </summary>
public class IniConfigFile : ConfigFile
{
    public IniConfigFile(string filePath, ConfigFileScope scope) : base(filePath, scope)
    {
    }

    protected override Dictionary<string, object> ReadSettings(string fileName)
    {
        var result = new Dictionary<string, object>();

        var exists = FileHelpers.FileExists(fileName);
        if (!exists) return result;

        try
        {
            var lines = File.ReadAllLines(fileName);
            foreach (var line in lines)
            {
                var trimmedLine = line.Trim();
                
                // Skip empty lines and comments
                if (string.IsNullOrWhiteSpace(trimmedLine) || trimmedLine.StartsWith("#") || trimmedLine.StartsWith(";"))
                {
                    continue;
                }

                // Parse KEY=VALUE format
                int equalsPos = trimmedLine.IndexOf('=');
                if (equalsPos > 0)
                {
                    var key = trimmedLine.Substring(0, equalsPos).Trim();
                    var value = trimmedLine.Substring(equalsPos + 1).Trim();

                    // Check if this is a list represented as comma-separated values
                    if (value.Contains(","))
                    {
                        // Convert comma-separated string to List<string>
                        var listValue = value.Split(',', StringSplitOptions.RemoveEmptyEntries)
                                            .Select(v => v.Trim())
                                            .ToList();
                        
                        // Convert from flat format to hierarchical format
                        var dotNotationKey = KnownSettings.ToDotNotation(key);
                        SetNestedValue(result, dotNotationKey.Split('.'), listValue);
                    }
                    else
                    {
                        // Convert from flat format to hierarchical format
                        var dotNotationKey = KnownSettings.ToDotNotation(key);
                        SetNestedValue(result, dotNotationKey.Split('.'), value);
                    }
                }
            }
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error reading INI configuration file: {ex.Message}");
        }

        return result;
    }

    protected override void WriteSettings(string fileName, Dictionary<string, object>? settings = null)
    {
        try
        {
            var flatData = FlattenDictionary(Settings);
            var lines = new StringBuilder();

            foreach (var pair in flatData)
            {
                var key = KnownSettings.ToEnvironmentVariable(pair.Key);
                var value = pair.Value?.ToString() ?? string.Empty;
                lines.AppendLine($"{key}={value}");
            }

            FileHelpers.WriteAllText(fileName, lines.ToString());
        }
        catch (Exception ex)
        {
            Console.Error.WriteLine($"Error writing INI configuration file: {ex.Message}");
        }
    }

    private void SetNestedValue(Dictionary<string, object> data, string[] keyParts, object value)
    {
        if (keyParts.Length == 1)
        {
            data[keyParts[0]] = value;
            return;
        }

        if (!data.ContainsKey(keyParts[0]))
        {
            data[keyParts[0]] = new Dictionary<string, object>();
        }

        if (data[keyParts[0]] is Dictionary<string, object> nestedDict)
        {
            SetNestedValue(nestedDict, keyParts[1..], value);
        }
        else
        {
            // Replace non-dictionary value with a dictionary
            var newDict = new Dictionary<string, object>();
            data[keyParts[0]] = newDict;
            SetNestedValue(newDict, keyParts[1..], value);
        }
    }

    private Dictionary<string, object> FlattenDictionary(Dictionary<string, object> data, string prefix = "")
    {
        var result = new Dictionary<string, object>();

        foreach (var pair in data)
        {
            var key = string.IsNullOrEmpty(prefix) ? pair.Key : $"{prefix}.{pair.Key}";

            if (pair.Value is Dictionary<string, object> nestedDict)
            {
                // Recursively flatten nested dictionaries
                foreach (var nestedPair in FlattenDictionary(nestedDict, key))
                {
                    result[nestedPair.Key] = nestedPair.Value;
                }
            }
            else if (pair.Value is List<string> stringList)
            {
                // Handle list values
                // In INI format, we'll just concatenate list items with a separator
                // This is a simplification - in a real implementation, you might want to handle this differently
                result[key] = string.Join(",", stringList);
            }
            else
            {
                result[key] = pair.Value;
            }
        }

        return result;
    }
}

File: src/commonjs/Configuration/IniConfigFile.ts
import * as fs from 'fs';
import { ConfigFile } from './ConfigFile';
import { ConfigFileScope } from './ConfigFileScope';
import { KnownSettings } from './KnownSettings';

/**
 * Handles reading and writing INI-style configuration files.
 */
export class IniConfigFile extends ConfigFile {
    public constructor(filePath: string, scope: ConfigFileScope) {
        super(filePath, scope);
    }

    protected ReadSettings(fileName: string): Map<string, any> {
        const result = new Map<string, any>();

        if (!fs.existsSync(fileName)) return result;

        try {
            const content = fs.readFileSync(fileName, 'utf-8');
            const lines = content.split('\n');
            
            for (const line of lines) {
                const trimmedLine = line.trim();
                
                // Skip empty lines and comments
                if (!trimmedLine || trimmedLine.startsWith('#') || trimmedLine.startsWith(';')) {
                    continue;
                }

                // Parse KEY=VALUE format
                const equalsPos = trimmedLine.indexOf('=');
                if (equalsPos > 0) {
                    const key = trimmedLine.substring(0, equalsPos).trim();
                    const value = trimmedLine.substring(equalsPos + 1).trim();

                    // Check if this is a list represented as comma-separated values
                    if (value.includes(',')) {
                        // Convert comma-separated string to string array
                        const listValue = value.split(',')
                                            .map(v => v.trim())
                                            .filter(v => v.length > 0);
                        
                        // Convert from flat format to hierarchical format
                        const dotNotationKey = KnownSettings.ToDotNotation(key);
                        this.SetNestedValue(result, dotNotationKey.split('.'), listValue);
                    } else {
                        // Convert from flat format to hierarchical format
                        const dotNotationKey = KnownSettings.ToDotNotation(key);
                        this.SetNestedValue(result, dotNotationKey.split('.'), value);
                    }
                }
            }
        } catch (ex: any) {
            console.error(`Error reading INI configuration file: ${ex.message}`);
        }

        return result;
    }

    protected WriteSettings(fileName: string, settings?: Map<string, any> | null): void {
        try {
            const flatData = this.FlattenDictionary(this.Settings);
            let content = '';

            for (const [key, value] of flatData) {
                const envKey = KnownSettings.ToEnvironmentVariable(key);
                const valueStr = value?.toString() ?? '';
                content += `${envKey}=${valueStr}\n`;
            }

            fs.writeFileSync(fileName, content);
        } catch (ex: any) {
            console.error(`Error writing INI configuration file: ${ex.message}`);
        }
    }

    private SetNestedValue(data: Map<string, any>, keyParts: string[], value: any): void {
        if (keyParts.length === 1) {
            data.set(keyParts[0], value);
            return;
        }

        if (!data.has(keyParts[0])) {
            data.set(keyParts[0], new Map<string, any>());
        }

        const nested = data.get(keyParts[0]);
        if (nested instanceof Map) {
            this.SetNestedValue(nested, keyParts.slice(1), value);
        } else {
            // Replace non-Map value with a Map
            const newMap = new Map<string, any>();
            data.set(keyParts[0], newMap);
            this.SetNestedValue(newMap, keyParts.slice(1), value);
        }
    }

    private FlattenDictionary(data: Map<string, any>, prefix: string = ''): Map<string, any> {
        const result = new Map<string, any>();

        for (const [key, value] of data) {
            const fullKey = prefix ? `${prefix}.${key}` : key;

            if (value instanceof Map) {
                // Recursively flatten nested Maps
                for (const [nestedKey, nestedValue] of this.FlattenDictionary(value, fullKey)) {
                    result.set(nestedKey, nestedValue);
                }
            } else if (Array.isArray(value)) {
                // Handle array values
                // In INI format, we'll just concatenate array items with a separator
                result.set(fullKey, value.join(','));
            } else {
                result.set(fullKey, value);
            }
        }

        return result;
    }
}